src/data_storage/merging.test.ts

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
+ * See, the License, for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Filename: merging.test.ts
+ * Description:
+ *   TODO
+ */
+
+import {testProp, fc} from 'jest-fast-check';
+import PouchDB from 'pouchdb';
+import {ProjectID} from './datamodel/core';
+import {Record} from './datamodel/ui';
+import {
+  deleteFAIMSDataForID,
+  generateFAIMSDataID,
+  getFirstRecordHead,
+  getFullRecordData,
+  listFAIMSProjectRevisions,
+  undeleteFAIMSDataForID,
+  upsertFAIMSData,
+} from './data_storage';
+import {equals} from './utils/eqTestSupport';
+
+import {getDataDB} from './sync/index';
+
+PouchDB.plugin(require('pouchdb-adapter-memory')); // enable memory adapter for testing
+
+const projdbs: any = {};
+
+function mockDataDB(project_id: ProjectID) {
+  if (projdbs[project_id] === undefined) {
+    const db = new PouchDB(project_id, {adapter: 'memory'});
+    projdbs[project_id] = db;
+  }
+  return projdbs[project_id];
+}
+
+async function cleanDataDBS() {
+  let db;
+  for (const project_id in projdbs) {
+    db = projdbs[project_id];
+    delete projdbs[project_id];
+
+    if (db !== undefined) {
+      try {
+        await db.destroy();
+        //await db.close();
+      } catch (err) {
+        console.error(err);
+      }
+    }
+  }
+}
+
+jest.mock('./sync/index', () => ({
+  getDataDB: mockDataDB,
+}));
+
+
+describe('test basic automerge', () => {
+  test('single revision', async () => {
+    // This tests the case where there is a single revision (i.e. new record)
+  });
+  test('no merged needed', async () => {
+    // This tests the case where there is a linear history, so there's no need
+    // for merging
+  });
+  test('extra head', async () => {
+    // This tests the case where there is a linear history, but where an old
+    // head was not removed (this shouldn't happen, but maybe there's some bad
+    // integration code that wrote to couchdb
+  });
+  test('same change', async () => {
+    // This tests the case where there has been a split, but the same change has
+    // been made
+  });
+  test('different change', async () => {
+    // This tests the case where there has been a split, and different changes
+    // have been made. This should cause the basic automerge to fail.
+  });
+  test('changes to different avps', async () => {
+    // This tests the case where there has been a split, but the changes have
+    // been to different avps
+  });
+  test('changes to different avps AND different change', async () => {
+    // This tests the case where there are three heads, of which two can be
+    // merged
+  });
+  test('changes to different avps AND different change 4 HEADS', async () => {
+    // This tests the case where there are 4 heads, of which three can be merged
```

Comment:
Just thinking about classical normalisation anomalies. If someone unchecks all items of a checkbox, that's still update, not delete in this context, right? (We should never be issuing a delete anyways). And I think derived values (I haven't heard any design discussion around this, so feel free to put this side of things off) will be derived at derivation-target-update, rather than at view, so to this system they just look like normal writes.

Ah, what about timetravel? Bad clocks are evil -- are we relying on timestamps for *literally anything?* as part of this process? I hope not. What if we have temporally distant updates?

Ex:

Device 1: 1. Make record -> 2. Sync -> 3. Go offline -> 18 or so. Edit record -> 19. Go Online 20. Sync
Device 2: 2. Sync, 4. Edit, 5. Sync, 6. Edit, 7. Sync. 8 Edit ... The fact that device 1's edit is so far back in the "history" of the thing... is that part of this testing regimen?

Reply:
Yep, changes to a form field would result in a new avp, so no deletion (though that brings up a case we need to check, what if you try to merge a revision which is marked as deleted, with a non-deleted one)?

Automerging doesn't care about timestamps, the only time timestamps (or time) comes up in the merging code is what time to set for the new revision (so nothing stops there being a merge revision where its parents have creation times after it because someone's local device clock is massively off, that's the users fault if they get themselves confused).

Note that in the case where there is a conflict the automerger *cannot* fix, and the *user* has to do a manual merge, if the user thinks that because one version has a newer timestamp (due to bad clocks) and goes with that one, that's not something I think we can handle (other than doing something really horrible like trying to correct timestamps in revisions given an offset computed for that bad device, which seems to me to be a path of madness).

Reply:
No, we are not expected to handle horrible timestamps. 2.6 had... a bad method for that, but it was needed. And yes, the case you raise is one we need to check for.